File: android/app/src/main/java/com/harshbshah/fer/ui/viewmodel/HistoryViewModel.kt
package com.harshbshah.fer.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.harshbshah.fer.data.ExerciseLibrary
import com.harshbshah.fer.data.model.MuscleGroup
import com.harshbshah.fer.data.model.SetEntry
import com.harshbshah.fer.data.model.WorkoutSession
import com.harshbshah.fer.data.repository.FirestoreRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import java.util.Calendar
import java.util.Date

class HistoryViewModel(private val repository: FirestoreRepository) : ViewModel() {
    private val _workouts = MutableStateFlow<List<WorkoutSession>>(emptyList())
    val workouts: StateFlow<List<WorkoutSession>> = _workouts

    init {
        viewModelScope.launch {
            repository.workoutsFlow().collect { _workouts.value = it }
        }
    }

    private fun startOfDay(date: Date): Date {
        val cal = Calendar.getInstance()
        cal.time = date
        cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
        cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
        return cal.time
    }

    val currentStreak: Int
        get() {
            val days = _workouts.value.map { startOfDay(it.startedAt) }.toSet()
            var streak = 0
            var cursor = startOfDay(Date())
            while (days.contains(cursor)) {
                streak++
                val cal = Calendar.getInstance(); cal.time = cursor; cal.add(Calendar.DAY_OF_YEAR, -1)
                cursor = cal.time
            }
            return streak
        }

    val workoutsThisWeek: Int
        get() {
            val cal = Calendar.getInstance(); cal.add(Calendar.DAY_OF_YEAR, -7)
            val weekAgo = cal.time
            return _workouts.value.count { it.startedAt >= weekAgo }
        }

    fun delete(workout: WorkoutSession) {
        val id = workout.id ?: return
        viewModelScope.launch { runCatching { repository.deleteWorkout(id) } }
    }

    fun history(exerciseId: String): List<Pair<Date, SetEntry>> =
        repository.history(exerciseId, _workouts.value)

    fun workoutDates(lastDays: Int): Set<Date> {
        val cal = Calendar.getInstance(); cal.add(Calendar.DAY_OF_YEAR, -(lastDays - 1))
        val cutoff = startOfDay(cal.time)
        return _workouts.value.mapNotNull { w ->
            val day = startOfDay(w.startedAt)
            if (day >= cutoff) day else null
        }.toSet()
    }

    fun dailyVolume(lastDays: Int): List<Pair<Date, Double>> {
        val today = startOfDay(Date())
        val volumeByDay = mutableMapOf<Date, Double>()
        for (w in _workouts.value) {
            val day = startOfDay(w.startedAt)
            volumeByDay[day] = (volumeByDay[day] ?: 0.0) + w.totalVolume
        }
        return (0 until lastDays).reversed().map { offset ->
            val cal = Calendar.getInstance(); cal.time = today; cal.add(Calendar.DAY_OF_YEAR, -offset)
            val day = cal.time
            day to (volumeByDay[day] ?: 0.0)
        }
    }

    fun weeklyVolume(weeks: Int): List<Pair<Date, Double>> {
        fun weekStart(date: Date): Date {
            val cal = Calendar.getInstance()
            cal.firstDayOfWeek = Calendar.MONDAY
            cal.time = date
            cal.set(Calendar.DAY_OF_WEEK, cal.firstDayOfWeek)
            cal.set(Calendar.HOUR_OF_DAY, 0); cal.set(Calendar.MINUTE, 0)
            cal.set(Calendar.SECOND, 0); cal.set(Calendar.MILLISECOND, 0)
            return cal.time
        }
        val thisWeekStart = weekStart(Date())
        val volumeByWeek = mutableMapOf<Date, Double>()
        for (w in _workouts.value) {
            val ws = weekStart(w.startedAt)
            volumeByWeek[ws] = (volumeByWeek[ws] ?: 0.0) + w.totalVolume
        }
        return (0 until weeks).reversed().map { offset ->
            val cal = Calendar.getInstance(); cal.time = thisWeekStart; cal.add(Calendar.WEEK_OF_YEAR, -offset)
            val ws = cal.time
            ws to (volumeByWeek[ws] ?: 0.0)
        }
    }

    fun volumeByMuscleGroup(): List<Pair<MuscleGroup, Double>> {
        val exercisesById = ExerciseLibrary.all.associateBy { it.id }
        val volumeByMuscle = mutableMapOf<MuscleGroup, Double>()
        for (workout in _workouts.value) {
            for (exercise in workout.exercises) {
                val muscle = exercisesById[exercise.exerciseId]?.primaryMuscle ?: continue
                val volume = exercise.sets.filter { it.isCompleted }.sumOf { it.weight * it.reps }
                volumeByMuscle[muscle] = (volumeByMuscle[muscle] ?: 0.0) + volume
            }
        }
        return volumeByMuscle.filter { it.value > 0 }.map { it.key to it.value }.sortedByDescending { it.second }
    }
}

File: Fer/ViewModels/HistoryViewModel.swift
//
//  HistoryViewModel.swift
//  Fer
//

import Foundation
import FirebaseFirestore
import Combine

@MainActor
final class HistoryViewModel: ObservableObject {
    @Published var workouts: [WorkoutSession] = []
    private var listener: ListenerRegistration?

    init() {
        listener = FirestoreService.shared.workoutsListener { [weak self] workouts in
            self?.workouts = workouts
        }
    }

    deinit {
        listener?.remove()
    }

    var currentStreak: Int {
        let calendar = Calendar.current
        let days = Set(workouts.map { calendar.startOfDay(for: $0.startedAt) })
        var streak = 0
        var cursor = calendar.startOfDay(for: Date())
        while days.contains(cursor) {
            streak += 1
            guard let prev = calendar.date(byAdding: .day, value: -1, to: cursor) else { break }
            cursor = prev
        }
        return streak
    }

    var workoutsThisWeek: Int {
        let calendar = Calendar.current
        guard let weekAgo = calendar.date(byAdding: .day, value: -7, to: Date()) else { return 0 }
        return workouts.filter { $0.startedAt >= weekAgo }.count
    }

    func delete(_ workout: WorkoutSession) {
        guard let id = workout.id else { return }
        Haptics.light()
        Task { try? await FirestoreService.shared.deleteWorkout(id: id) }
    }

    func history(forExerciseId id: String) -> [(date: Date, bestSet: SetEntry)] {
        FirestoreService.shared.history(forExerciseId: id, workouts: workouts)
    }

    /// Start-of-day dates (in the last `days` days, including today) that had at least one workout.
    func workoutDates(last days: Int) -> Set<Date> {
        let calendar = Calendar.current
        guard let cutoff = calendar.date(byAdding: .day, value: -(days - 1), to: calendar.startOfDay(for: Date())) else {
            return []
        }
        return Set(workouts.compactMap { workout -> Date? in
            let day = calendar.startOfDay(for: workout.startedAt)
            return day >= cutoff ? day : nil
        })
    }

    /// Total volume per day for the last `days` days, oldest first, zero-filled for rest days.
    func dailyVolume(last days: Int) -> [(date: Date, volume: Double)] {
        let calendar = Calendar.current
        let today = calendar.startOfDay(for: Date())
        var volumeByDay: [Date: Double] = [:]
        for workout in workouts {
            let day = calendar.startOfDay(for: workout.startedAt)
            volumeByDay[day, default: 0] += workout.totalVolume
        }
        return (0..<days).reversed().compactMap { offset in
            guard let day = calendar.date(byAdding: .day, value: -offset, to: today) else { return nil }
            return (date: day, volume: volumeByDay[day] ?? 0)
        }
    }

    /// Total volume per week for the last `weeks` weeks, oldest first.
    func weeklyVolume(weeks: Int) -> [(weekStart: Date, volume: Double)] {
        let calendar = Calendar.current
        guard let thisWeekStart = calendar.date(from: calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: Date())) else {
            return []
        }
        var volumeByWeek: [Date: Double] = [:]
        for workout in workouts {
            guard let weekStart = calendar.date(from: calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], from: workout.startedAt)) else { continue }
            volumeByWeek[weekStart, default: 0] += workout.totalVolume
        }
        return (0..<weeks).reversed().compactMap { offset in
            guard let weekStart = calendar.date(byAdding: .weekOfYear, value: -offset, to: thisWeekStart) else { return nil }
            return (weekStart: weekStart, volume: volumeByWeek[weekStart] ?? 0)
        }
    }

    /// Total completed-set volume grouped by primary muscle group, largest first.
    func volumeByMuscleGroup() -> [(muscle: MuscleGroup, volume: Double)] {
        var volumeByMuscle: [MuscleGroup: Double] = [:]
        let exercisesById = Dictionary(uniqueKeysWithValues: ExerciseLibrary.all.map { ($0.id, $0) })
        for workout in workouts {
            for exercise in workout.exercises {
                guard let muscle = exercisesById[exercise.exerciseId]?.primaryMuscle else { continue }
                let volume = exercise.sets.filter(\.isCompleted).reduce(0) { $0 + ($1.weight * Double($1.reps)) }
                volumeByMuscle[muscle, default: 0] += volume
            }
        }
        return volumeByMuscle
            .filter { $0.value > 0 }
            .map { (muscle: $0.key, volume: $0.value) }
            .sorted { $0.volume > $1.volume }
    }
}
